File: ua-agent/frontend/src/shared/contracts/capture.ts
import { z } from "zod";

export const PostIdSource = z.enum([
  "aweme_id",
  "share_url_canonical",
  "share_url_short",
  "xhs_note_url",
]);
export type PostIdSource = z.infer<typeof PostIdSource>;

export const NoteType = z.enum(["video", "image_text"]);
export type NoteType = z.infer<typeof NoteType>;

/** Platform tag for multi-platform keyword crawl (006). */
export const Platform = z.enum(["douyin", "xiaohongshu"]);
export type Platform = z.infer<typeof Platform>;

/** Media form of a captured material item. */
export const MediaKind = z.enum(["video", "images", "mixed"]);
export type MediaKind = z.infer<typeof MediaKind>;

export const commentItemSchema = z.object({
  author: z.string().max(256),
  content: z.string().max(1024),
  like_count: z.number().int().gte(-1),
  time_text: z.string().max(64),
});
export type CommentItem = z.infer<typeof commentItemSchema>;

export const materialEntrySchema = z.object({
  post_id: z.string().min(1).max(128),
  post_id_source: PostIdSource,
  share_url: z.string().min(1).max(2048),
  share_text: z.string().min(1).max(4096),
  caption: z.string().max(4096),
  author_handle: z.string().min(1).max(256),
  author_display_name: z.union([z.string().max(256), z.null()]),
  hashtags: z.array(z.string().max(64)).max(64),
  music_id: z.union([z.string().max(128), z.null()]),
  music_title: z.union([z.string().max(256), z.null()]),
  like_count: z.number().int().gte(-1),
  comment_count: z.number().int().gte(-1),
  share_count: z.number().int().gte(-1),
  collect_count: z.number().int().gte(-1),
  author_follower_count: z.union([z.number().int().positive(), z.null()]),
  captured_at: z.string().datetime({ offset: false }),
  captured_by_device: z.string().min(1).max(128),
  note_type: NoteType.default("video"),
  platform: Platform.default("douyin"),
  media_kind: MediaKind.default("video"),
  image_urls: z.union([z.array(z.string().url().max(2048)).max(64), z.null()]).default(null),
  comments: z.array(commentItemSchema).max(10).default([]),
  transcript: z.union([z.string().max(65536), z.null()]).default(null),
  transcribed_at: z.union([z.string().datetime({ offset: false }), z.null()]).default(null),
});
export type MaterialEntry = z.infer<typeof materialEntrySchema>;
export const MaterialEntry = materialEntrySchema;

File: ua-agent/backend/src/ua_agent/contracts/capture.py
from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

PostIdSource = Literal[
    "aweme_id",
    "share_url_canonical",
    "share_url_short",
    "xhs_note_url",
]
NoteType = Literal["video", "image_text"]


class CommentItem(BaseModel):
    """A single top-level comment captured from the XHS detail overlay."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    author: Annotated[str, Field(max_length=256)]
    content: Annotated[str, Field(max_length=1024)]
    like_count: Annotated[int, Field(ge=-1)]
    time_text: Annotated[str, Field(max_length=64)]


class MaterialEntry(BaseModel):
    """A single saved piece of Douyin content. Both wire shape and SQLite row.

    `like_count`, `comment_count`, `share_count`, `collect_count` use -1 as the
    "not visible at capture time" sentinel; 0 means "visible and zero".
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    post_id: Annotated[str, Field(min_length=1, max_length=128)]
    post_id_source: PostIdSource
    share_url: Annotated[str, Field(min_length=1, max_length=2048)]
    share_text: Annotated[str, Field(min_length=1, max_length=4096)]
    caption: Annotated[str, Field(max_length=4096)]
    author_handle: Annotated[str, Field(min_length=1, max_length=256)]
    author_display_name: Annotated[str | None, Field(max_length=256)]
    hashtags: Annotated[list[Annotated[str, Field(max_length=64)]], Field(max_length=64)]
    music_id: Annotated[str | None, Field(max_length=128)]
    music_title: Annotated[str | None, Field(max_length=256)]
    like_count: Annotated[int, Field(ge=-1)]
    comment_count: Annotated[int, Field(ge=-1)]
    share_count: Annotated[int, Field(ge=-1)]
    collect_count: Annotated[int, Field(ge=-1)]
    captured_at: Annotated[
        str,
        Field(
            description=(
                "RFC 3339 UTC, ms precision, trailing Z. "
                "Example: 2026-05-02T12:34:56.789Z."
            ),
        ),
    ]
    captured_by_device: Annotated[str, Field(min_length=1, max_length=128)]
    note_type: NoteType = "video"
    comments: list[CommentItem] = []
